python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+  result = await client.execute_workflow("MyWorkflow", "arg1", id="someid", task_queue="somequeue")
+  print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+def say_hello(name: str) -> str:
+  return f"Hello, {name}!"
+
+def main(client: temporalio.Client):
+  worker = temporalio.Worker(client, task_queue="somequeue")
+  worker.register_activity(say_hello)
+  await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(
+        self,
+        addr: str,
+        *,
+        namespace: str = "default",
+        default_task_queue: str = None,
+        identity: str = None
+        # ... additional options omitted for brevity
+    ) -> None:
+        # Actually dials gRPC here
+        pass
+
+    # Raw gRPC service
+    @property
+    def service(self) -> workflowservice.WorkflowService:
+        pass
+
+    def start_workflow(
+        workflow: str,
+        *args,
+        id: str = None,
+        task_queue: str = None,
+        retry_policy: RetryPolicy = None,
+        # Note, we are embedding signal-with-start in here
+        start_signal: str = None,
+        start_signal_args=[],
+        # ... additional options omitted for brevity
+    ) -> WorkflowHandle[Any]:
+        pass
+
+    # In a future phase:
+    # @overload
+    # def start_workflow(workflow: Callable[P, T], *args: P.args, ...) -> WorkflowExecution[T]
+
+    # Does not validate whether a workflow exists
+    def workflow(
+        workflow_id: str,
+        run_id: str = None
+        # ...
+    ) -> WorkflowHandle[Any]:
+        pass
+
+
+class WorkflowHandle(Generic[T]):
+    def __init__(
+        self,
+        client: Client,
+        workflow_id: str,
+        *,
+        run_id: str = None
+        # ...
+    ) -> None:
+        pass
+
+    def __await__(self) -> T:
+        pass
+
+    async def cancel(self):
+        pass
+
+    async def describe(self):
+        pass
+
+    async def query(self, name: str, *args):
+        pass
+
+    async def signal(self, name: str, *args):
+        pass
+
+    async def terminate(self, *, reason: str = None):
+        pass
+```
+
+Notes:
+
+* We are intentionally defining a `Client` instead of a `WorkflowClient` since this will be general purpose
```

Comment:
Let's discuss this because I'm not convinced this is the right approach.
What other general purpose methods would this client have?

Reply:
>  What other general purpose methods would this client have?

Off the top of my head probably only async activity stuff. From below:

> but at least for now I can't see defining any more than the two methods and the one property

It is a good, simple concept to "create client, call client". We are trying to hard to add words and terms and steps and other barriers to simplicity.

Reply:
Per other conversation, we're good here